lib/ansible/modules/web_infrastructure/ansible_tower/tower_inventory_source.py

```diff
@@ -0,0 +1,277 @@
+        update_on_launch=dict(type='bool', required=False),
+        update_cache_timeout=dict(type='int', required=False),
+        state=dict(choices=['present', 'absent'], default='present'),
+    ))
+
+    module = AnsibleModule(argument_spec=argument_spec,
+                           supports_check_mode=True)
+
+    if not HAS_TOWER_CLI:
+        module.fail_json(msg='ansible-tower-cli required for this module')
+
+    name = module.params.get('name')
+    inventory = module.params.get('inventory')
+    source = module.params.get('source')
+    state = module.params.get('state')
+
+    json_output = {'inventory_source': name, 'state': state}
+
+    tower_auth = tower_auth_config(module)
+    with settings.runtime_values(**tower_auth):
+        tower_check_mode(module)
+        inventory_source = tower_cli.get_resource('inventory_source')
+
+        try:
+            params = {}
+            params['create_on_missing'] = True
+            params['name'] = name
+            params['source'] = source
+
+            if module.params.get('description'):
+                params['description'] = module.params.get('description')
+
+            if module.params.get('credential'):
+                credential_res = tower_cli.get_resource('credential')
+                credential = credential_res.get(
+                    name=module.params.get('credential'))
+                params['credential'] = credential['id']
+
+            if module.params.get('source_project'):
+                source_project_res = tower_cli.get_resource('source_project')
+                source_project = source_project_res.get(
+                    name=module.params.get('source_project'))
+                params['source_project'] = source_project['id']
+
+            if module.params.get('source_script'):
+                source_script_res = tower_cli.get_resource('inventory_script')
+                script = source_script_res.get(
+                    name=module.params.get('source_script'))
+                params['source_script'] = script['id']
+
+            inventory_res = tower_cli.get_resource('inventory')
+            params['inventory'] = inventory_res.get(name=inventory)['id']
+
+            for key in ('source_vars', 'timeout', 'source_path',
+                        'update_on_project_update', 'source_regions',
+                        'instance_filters', 'group_by', 'overwrite',
+                        'overwrite_vars', 'update_on_launch',
+                        'update_cache_timeout'):
+                if 'source' in params:
+                    params[key] = module.params.get(key)
+                elif module.params.get(key):
+                    params.setdefault('inputs', {})[key] = module.params.get(key)
+
+            if state == 'present':
+                result = inventory_source.modify(**params)
+                json_output['id'] = result['id']
+            elif state == 'absent':
+                result = inventory_source.delete(**params)
```

Comment:
for delete, you might want to consider the `fail_on_missing`, because I'm guessing that you want it to not fail on missing (OK, changed=False).

Reply:
Thanks for spotting that. I've added `fail_on_missing`, I thought it was the base behavior.